### PR TITLE
[lldb] Fix clang importer for swift caching debugging (attempt #2)

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Further improve how lldb constructing clang importer to utilize explicit modules:
* Add logics to initialize the swift ASTContext from direct cc1 arguments embedded in the swiftmodule
* Make sure the direct cc1 arguments are not repeated constructed. It should only be constructed as a fresh invocation
* Make sure the lldb doesn't overwrite the explicit module build settings when filter arguments. This ensures the already built explicit modules are used.
* Use the newly added clang option that ignore CAS info when loading PCMs. This allows lldb to load CAS enabled PCM directly from disk.

rdar://135611011

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 1]
```diff
@@ -1764,23 +1751,37 @@ static void applyOverrideOptions(std::vector<std::string> &args,
...
   // which are very specific to one compiler version and cannot
   // be merged with driver options.
   bool fresh_invocation = importer_options.ExtraArgs.empty();
-  if (fresh_invocation && !ExtraArgs.empty() && ExtraArgs.front() == "-cc1")
-    importer_options.DirectClangCC1ModuleBuild = true;
-  if (!importer_options.DirectClangCC1ModuleBuild && !ExtraArgs.empty() &&
-      ExtraArgs.front() == "-cc1")
+  bool invocation_direct_cc1 = ExtraArgs.front() == "-cc1";
+
+  // If it is not a fresh invocation, make sure the cc1 option match.
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp]:
```suggestion
  // If it is not a fresh invocation, make sure the cc1 option matches.
```

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 2]
```diff
@@ -1764,23 +1751,37 @@ static void applyOverrideOptions(std::vector<std::string> &args,
...
     AddDiagnostic(
         eSeverityWarning,
         "Mixing and matching of driver and cc1 Clang options detected");
 
-  AddExtraClangArgs(ExtraArgs, importer_options.ExtraArgs,
-                    importer_options.DirectClangCC1ModuleBuild);
+  importer_options.DirectClangCC1ModuleBuild = invocation_direct_cc1;
+
+  // If using direct cc1 flags, compute the arguments and return.
+  // Since this is cc1 flags, no driver overwrite can be applied.
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp]: What is a "driver overwrite"?

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 3]
```diff
@@ -1792,6 +1793,76 @@ void SwiftASTContext::AddExtraClangArgs(
...
+
+  // Ignore CAS info inside modules when loading.
+  invocation.getFrontendOpts().ModuleLoadIgnoreCAS = true;
+
+  // Remove non-existing modules in a systematic way.
+  bool module_missing = false;
+  auto CheckFileExists = [&](const char *file) {
+    if (!llvm::sys::fs::exists(file)) {
+      std::string m_description;
+      HEALTH_LOG_PRINTF("Nonexistent explicit module file %s", file);
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp]: I think we should instead call AddDiagnostic() here. This way it gets logged, and it's reported as an error to the user as the result of their command.

File: lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp [hunk 4]
```diff
@@ -1792,6 +1793,76 @@ void SwiftASTContext::AddExtraClangArgs(
...
+      module_missing = true;
+    }
+  };
+  llvm::for_each(invocation.getHeaderSearchOpts().PrebuiltModuleFiles,
+                 [&](const auto &mod) { CheckFileExists(mod.second.c_str()); });
+  llvm::for_each(invocation.getFrontendOpts().ModuleFiles,
+                 [&](const auto &mod) { CheckFileExists(mod.c_str()); });
+
+  // If missing, clear all the prebuilt module options and use implicit module
```
Comment (adrian-prantl) [inline lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp]:
```suggestion
  // If missing, clear all the prebuilt module options and switch to implicit modules.
```

**Review by adrian-prantl — COMMENTED** [review]
The code and tests all look good make sense. I have one suggestion inline about the error reporting.

**Comment by adrian-prantl** [comment]
Oh, and you probably want to create this PR against the `rebranch` scheme instead.

**Comment by cachemeifyoucan** [comment]
> Oh, and you probably want to create this PR against the `rebranch` scheme instead.

It doesn't hurt to commit to `main` related branch right? I can get full testing here. I will just cherry-pick to other branches after it is cleared here.

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test windows platform

**Comment by cachemeifyoucan** [comment]
@swift-ci please test windows platform

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Review by adrian-prantl — APPROVED** [review]

**Pull request merged**